onnx/defs/data_propagators.h

```diff
@@ -16,4 +49,160 @@ inline void PropagateShapeDataFromInputToOutput(DataPropagationContext& ctx, int
   }
 }
 
+// Data propagation function for Shape op
+// Propagates input shape to output shape
+inline void ShapeOpDataPropagator(DataPropagationContext& ctx) {
+  if (!hasNInputShapes(ctx, 1)) {
+    return;
+  }
+  if (ctx.getInputType(0)->tensor_type().has_shape()) {
+    auto input_shape = ctx.getInputType(0)->tensor_type().shape();
+    TensorShapeProto tsp;
+    tsp.CopyFrom(input_shape);
+    ctx.addOutputData(0, std::move(tsp));
+  }
+}
+
+inline void SizeOpDataPropagator(DataPropagationContext& ctx) {
+  const auto input_data = ctx.getInputData(0);
+  if (input_data != nullptr) {
+    TensorShapeProto tsp;
+    tsp.mutable_dim()->Add()->set_dim_value(input_data->dim_size());
+    ctx.addOutputData(0, std::move(tsp));
+  }
+}
+
+inline int MathOpTwoIntegers(std::string op_type, int a, int b) {
+  if (op_type == "Add") {
+    return a + b;
+  } else if (op_type == "Sub") {
+    return a - b;
+  } else if (op_type == "Mul") {
+    return a * b;
+  }
+  fail_shape_inference("Wrong op_type name for running propagation: ", op_type);
+}
+
+inline void MathOpDataPropagator(DataPropagationContext& ctx, std::string op_type) {
+  const auto input_0 = ctx.getInputData(0);
+  const auto input_1 = ctx.getInputData(1);
+  if (input_0 == nullptr || input_1 == nullptr) {
+      return;
+  }
+  TensorShapeProto tsp;
+  if (input_0->dim_size() < input_1->dim_size()) {
```

Comment:
are you supporting broadcasting? I suggest for simplicity we only support dim_size == 1... the dim size of shape will always be 1 .i.e is 1D tensor

Reply:
After offline discussion, the minor broadcasting here should be valid. I have further considered failed broadcasting if the dim sizes are incompatible